File: julia/src/Clustering.jl
module Clustering

export Cluster, cluster_flows

using ..Candidates: CandidateIndex

struct Cluster
    flow_ids::Vector{Int}
    landing_cells::Vector{Tuple{Int, Int}}
end

function cluster_flows(candidates::CandidateIndex)
    cell_to_flows = Dict{Tuple{Int, Int}, Vector{Int}}()
    for (fid, fc) in candidates.flows
        for cell in keys(fc.slots)
            push!(get!(cell_to_flows, cell, Int[]), fid)
        end
    end

    visited = Set{Int}()
    clusters = Vector{Cluster}()

    for fid in keys(candidates.flows)
        if fid in visited
            continue
        end
        queue = [fid]
        idx = 1
        flow_ids = Int[]
        landing_cells = Set{Tuple{Int, Int}}()

        while idx <= length(queue)
            current = queue[idx]
            idx += 1
            if current in visited
                continue
            end
            push!(visited, current)
            push!(flow_ids, current)

            fc = candidates.flows[current]
            for cell in keys(fc.slots)
                push!(landing_cells, cell)
                for neighbour in get(cell_to_flows, cell, Int[])
                    if neighbour ∉ visited
                        push!(queue, neighbour)
                    end
                end
            end
        end

        sort!(flow_ids)
        lc_vec = sort(collect(landing_cells); by = x -> (x[1], x[2]))
        push!(clusters, Cluster(flow_ids, lc_vec))
    end

    sort!(clusters; by = c -> length(c.flow_ids), rev = true)
    return clusters
end

end # module

File: src/clustering.py
"""Determine independent flow clusters based on overlapping landing zones."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Set, Tuple

from .candidates import CandidateIndex


@dataclass(slots=True)
class Cluster:
    flow_ids: List[int]
    landing_cells: List[Tuple[int, int]]


def cluster_flows(index: CandidateIndex) -> List[Cluster]:
    cell_to_flows: Dict[Tuple[int, int], Set[int]] = {}
    for flow_id, flow_candidate in index.flows.items():
        for cell in flow_candidate.time_slots:
            cell_to_flows.setdefault(cell, set()).add(flow_id)

    visited: Set[int] = set()
    clusters: List[Cluster] = []

    for flow_id in index.flows:
        if flow_id in visited:
            continue

        queue = [flow_id]
        pointer = 0
        component_flows: Set[int] = set()
        component_cells: Set[Tuple[int, int]] = set()

        while pointer < len(queue):
            current = queue[pointer]
            pointer += 1
            if current in visited:
                continue
            visited.add(current)
            component_flows.add(current)
            candidate = index.flows[current]
            for cell in candidate.time_slots:
                component_cells.add(cell)
                for neighbor in cell_to_flows.get(cell, set()):
                    if neighbor not in visited:
                        queue.append(neighbor)

        clusters.append(
            Cluster(
                flow_ids=sorted(component_flows),
                landing_cells=sorted(component_cells),
            )
        )

    clusters.sort(key=lambda c: len(c.flow_ids), reverse=True)
    return clusters
